lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -377,6 +377,24 @@ defmodule Mix.Compilers.Elixir do
     end
   end
 
+  defp warning_diagnostics(sources) do
+    Enum.flat_map(sources, fn source(source: source, warnings: warnings) ->
+      for {line, message} <- warnings do
+        diagnostic({source, line, message}, :warning)
+      end
+    end)
+  end
+
+  defp diagnostic({file, line, message}, severity) do
+    %Mix.Task.Compiler.Diagnostic{
+      file: Path.absname(file),
```

Comment:
It feels we are abusing `Path.absname(file)`. We already call `Path.absname` in the compiler for all returned errors. Then we call it here again. We need to be clearer on what we expect from those paths. If we are already calling it here, we should probably remove those from the compiler.

Reply:
True -- the reason for this (and the `to_string(message)` also) is that the `each_warning` callback sometimes uses relative paths and always uses charlist messages. The errors and warnings returned from the compiler, however, are always absolute paths and string messages. I could refactor it to only use those, now that we're returning them. This would let us get rid of the `each_warning` callback, since it would be unnecessary after that.

Reply:
@JakeBecker Yup, each_warning and all_warnings feel redundant now, so :+1: if we can get rid of each_warning. If we can't, then we should change its argument to be a string too. Please do those changes in a separate commit in case we change our minds! Thank you! :heart: